### PR TITLE
[DataLoader] Disable the usage of shared memory when persistent_workers=False on single GPU

### LINKED ISSUE #4334: Bus error(core dumped) in EdgeDataLoader of unsupervised graphsage example (20M edges)
## To Reproduce

1. In the example code: https://github.com/dmlc/dgl/blob/master/examples/pytorch/graphsage/advanced/train_sampling_unsupervised.py
2. Feed a dataset with 10M nodes and 20M edges
3. Bus error(core dumped)  happened in EdgeDataLoader before training and DataLoader in inference.

## Environment

 - DGL Version (e.g., 1.0): 0.8.1cu101
 - Backend Library & Version (e.g., PyTorch 0.4.1, MXNet/Gluon 1.3): PyTorch 1.8.1+cu101
 - OS (e.g., Linux): Docker
 - How you installed DGL (`conda`, `pip`, source): pip
 - Build command you used (if compiling from source):
 - Python version: 3.6.13
 - CUDA/cuDNN version (if applicable): 101
 - GPU models and configuration (e.g. V100): V100
 - Any other relevant information:

### PR DESCRIPTION
**Pull request by BarclayII**
Should fix #4334 .

### CODE REVIEW

**Comment by Rhett-Ying** [comment]
Commit ID: 4ddb9ca24913edd93b38be60a787c9db4fa7e7a2

Build ID: 5

Status: ✅ CI test succeeded 

Report path: [link](https://dgl-ci-result.s3.us-west-2.amazonaws.com/dgl/PR-4497/5/5/logs/report.html) 

Full logs path: [link](https://dgl-ci-result.s3.us-west-2.amazonaws.com/dgl/PR-4497/5/5/logs/cireport.log)

**Comment by Rhett-Ying** [comment]
Commit ID: e3035ff0e050af32a21c8049bcba41b4a1cc71b5

Build ID: 6

Status: ✅ CI test succeeded 

Report path: [link](https://dgl-ci-result.s3.us-west-2.amazonaws.com/dgl/PR-4497/6/6/logs/report.html) 

Full logs path: [link](https://dgl-ci-result.s3.us-west-2.amazonaws.com/dgl/PR-4497/6/6/logs/cireport.log)

File: python/dgl/dataloading/dataloader.py [hunk 1]
```diff
@@ -147,10 +147,13 @@ def __init__(self, indices, batch_size, drop_last, shuffle):
...
         # the worker processes can see it when persistent_workers=True, where self._indices
         # would not be duplicated every epoch.
-        self._indices = torch.arange(self._id_tensor.shape[0], dtype=torch.int64).share_memory_()
+        self._indices = torch.arange(self._id_tensor.shape[0], dtype=torch.int64)
+        if use_shared_memory:
+            self._indices.share_memory_()
         self.batch_size = batch_size
         self.drop_last = drop_last
         self._shuffle = shuffle
+        self._indices.share_memory_()
```
Comment (yaox12) [inline python/dgl/dataloading/dataloader.py:156]: Redundant?

Comment (BarclayII) [reply python/dgl/dataloading/dataloader.py]: Oops.

**Review by yaox12 — APPROVED** [review]

**Pull request merged**